presto-array/src/main/java/com/facebook/presto/array/ReferenceCountMap.java

```diff
@@ -59,19 +62,8 @@ public long sizeOf()
         return INSTANCE_SIZE + SizeOf.sizeOf(key) + SizeOf.sizeOf(value) + SizeOf.sizeOf(used);
     }
 
-    private static final class ObjectStrategy
-            implements Strategy<Object>
+    private static long getHashCode(Object key, int size)
     {
-        @Override
-        public int hashCode(Object object)
-        {
-            return System.identityHashCode(object);
-        }
-
-        @Override
-        public boolean equals(Object left, Object right)
-        {
-            return left == right;
-        }
+        return (((long) System.identityHashCode(key)) << Integer.SIZE) + size;
```

Comment:
Do we really need the size? What happens if we use just identity hashcode for the hash? That's what we had before (and what Object-based maps use) and it was sufficient, no?

Reply:
Unfortunately, that's not sufficient. identityHashMap also needs to use `==` compare two objects to make sure they are the same. The collision can happen at about 2.5% out of 100 million objects. With different sizes however, the collision can be lower than 1% if there are 50 different sizes that are normally distributed. The collision can be lower than 0.01% if there are 1000 different sizes that are normally distributed. Of course, the experiment with sizes are quite bit artificial.